### PR TITLE
Share code between Heap ByteBuf implementations

### PR DESCRIPTION
**Pull request by normanmaurer**
Motiviation:

We have a lot of duplicated code which makes it hard to maintain.

Modification:

Move shared code to HeapByteBufUtil and use it in the implementations.

Result:

Less duplicated code and so easier to maintain.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler PTAL

**Comment by nmittler** [comment]
LGTM

File: buffer/src/main/java/io/netty/buffer/PooledHeapByteBuf.java [hunk 1]
```diff
@@ -151,44 +135,27 @@ public final int readBytes(GatheringByteChannel out, int length) throws IOExcept
...
     protected void _setMedium(int index, int   value) {
-        index = idx(index);
-        memory[index]     = (byte) (value >>> 16);
-        memory[index + 1] = (byte) (value >>> 8);
-        memory[index + 2] = (byte) value;
+        HeapByteBufUtil.setMedium(memory, idx(index), value);
     }
 
     @Override
     protected void _setInt(int index, int   value) {
```
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/PooledHeapByteBuf.java]: is there a reason for the spacing between `int     value`?

**Comment by Scottmitch** [comment]
@normanmaurer - just 1 nit (which may apply to a few spots) and lgtm

File: buffer/src/main/java/io/netty/buffer/PooledHeapByteBuf.java (hunk 1 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PooledHeapByteBuf.java]: nope... fat fingers. Let me fix it and pull in :)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (1055c8787e0e971cedf23767797079db1dd82e7a) , 4.1 (992f8db0b4559216a52b26a339f4f7f823807a4b) and master (913337111c82f966ade7d733ee60052df7f32f7e)

**Pull request closed without merging**